File: backend/app/routers/admin/media.py
from fastapi import APIRouter, Depends, HTTPException, UploadFile, File
from bson import ObjectId
from app.database import get_database
from app.middleware.auth import require_admin
from app.schemas.media import MediaResponse, ReorderMediaRequest
from app.models.media import new_media
from app.services.image_processing import process_and_upload_image
from app.services.s3 import generate_presigned_download_url, delete_file_from_s3

router = APIRouter()


@router.get("/{project_id}/media")
async def list_media(project_id: str, admin: dict = Depends(require_admin)):
    db = get_database()
    cursor = db.media.find({"project_id": project_id}).sort("sort_order", 1)
    items = []
    async for m in cursor:
        items.append(MediaResponse(
            id=str(m["_id"]),
            project_id=m["project_id"],
            original_url=generate_presigned_download_url(m["original_key"]),
            compressed_url=generate_presigned_download_url(m["compressed_key"]),
            thumbnail_url=generate_presigned_download_url(m["thumbnail_key"]),
            filename=m["filename"],
            mime_type=m["mime_type"],
            width=m["width"],
            height=m["height"],
            size_bytes=m["size_bytes"],
            compressed_size_bytes=m["compressed_size_bytes"],
            sort_order=m["sort_order"],
            uploaded_at=m["uploaded_at"],
            is_selected=m["is_selected"],
        ))
    return {"media": items}


@router.post("/{project_id}/media", status_code=201)
async def upload_media(
    project_id: str,
    files: list[UploadFile] = File(...),
    admin: dict = Depends(require_admin),
):
    db = get_database()
    project = await db.projects.find_one({"_id": ObjectId(project_id)})
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    count = await db.media.count_documents({"project_id": project_id})
    uploaded = []
    for idx, file in enumerate(files):
        file_bytes = await file.read()
        result = process_and_upload_image(file_bytes, project_id, file.filename, file.content_type)
        media_doc = new_media(
            project_id=project_id,
            original_key=result["original_key"],
            compressed_key=result["compressed_key"],
            thumbnail_key=result["thumbnail_key"],
            filename=file.filename,
            mime_type=file.content_type,
            width=result["width"],
            height=result["height"],
            size_bytes=result["size_bytes"],
            compressed_size_bytes=result["compressed_size_bytes"],
            sort_order=count + idx,
        )
        insert = await db.media.insert_one(media_doc)
        uploaded.append(str(insert.inserted_id))
    return {"media_ids": uploaded, "message": f"{len(uploaded)} files uploaded"}


@router.delete("/{project_id}/media/{media_id}")
async def delete_media(project_id: str, media_id: str, admin: dict = Depends(require_admin)):
    db = get_database()
    m = await db.media.find_one({"_id": ObjectId(media_id), "project_id": project_id})
    if not m:
        raise HTTPException(status_code=404, detail="Media not found")
    delete_file_from_s3(m["original_key"])
    delete_file_from_s3(m["compressed_key"])
    delete_file_from_s3(m["thumbnail_key"])
    await db.media.delete_one({"_id": ObjectId(media_id)})
    return {"message": "Deleted"}


@router.put("/{project_id}/media/reorder")
async def reorder_media(project_id: str, body: ReorderMediaRequest, admin: dict = Depends(require_admin)):
    db = get_database()
    for idx, media_id in enumerate(body.media_ids):
        await db.media.update_one(
            {"_id": ObjectId(media_id), "project_id": project_id},
            {"$set": {"sort_order": idx}},
        )
    return {"message": "Reordered"}

File: src/routes/admin/media.ts
import { Router, Response } from 'express';
import { ObjectId } from 'mongodb';
import multer from 'multer';
import { getDb } from '../../database';
import { requireAdmin, AuthRequest } from '../../middleware/auth';
import { newMedia } from '../../models/media';
import { processAndUploadImage } from '../../services/imageProcessing';
import { generatePresignedDownloadUrl, deleteFileFromS3 } from '../../services/s3';

const router = Router();
const upload = multer({ storage: multer.memoryStorage(), limits: { fileSize: 50 * 1024 * 1024 } });

router.get('/:projectId/media', requireAdmin, async (req: AuthRequest, res: Response): Promise<void> => {
  const db = await getDb();
  const mediaItems = await db.collection('media').find({ project_id: req.params.projectId }).sort({ sort_order: 1 }).toArray();
  const mapped = await Promise.all(mediaItems.map(async (m) => ({
    id: m._id.toString(),
    project_id: m.project_id,
    original_url: await generatePresignedDownloadUrl(m.original_key),
    compressed_url: await generatePresignedDownloadUrl(m.compressed_key),
    thumbnail_url: await generatePresignedDownloadUrl(m.thumbnail_key),
    filename: m.filename,
    mime_type: m.mime_type,
    width: m.width,
    height: m.height,
    size_bytes: m.size_bytes,
    compressed_size_bytes: m.compressed_size_bytes,
    sort_order: m.sort_order,
    uploaded_at: m.uploaded_at,
    is_selected: m.is_selected,
  })));
  res.json({ media: mapped });
});

router.post('/:projectId/media', requireAdmin, upload.array('files'), async (req: AuthRequest, res: Response): Promise<void> => {
  const files = req.files as Express.Multer.File[] | undefined;
  if (!files?.length) { res.status(400).json({ detail: 'No files uploaded' }); return; }
  const db = await getDb();
  const project = await db.collection('projects').findOne({ _id: new ObjectId(req.params.projectId) });
  if (!project) { res.status(404).json({ detail: 'Project not found' }); return; }
  const count = await db.collection('media').countDocuments({ project_id: req.params.projectId });
  const uploadedIds: string[] = [];
  for (let idx = 0; idx < files.length; idx++) {
    const file = files[idx];
    const result = await processAndUploadImage(file.buffer, req.params.projectId, file.originalname, file.mimetype);
    const mediaDoc = newMedia(req.params.projectId, result.original_key, result.compressed_key, result.thumbnail_key, result.watermarked_key, file.originalname, file.mimetype, result.width, result.height, result.size_bytes, result.compressed_size_bytes, count + idx);
    const insert = await db.collection('media').insertOne(mediaDoc);
    uploadedIds.push(insert.insertedId.toString());
  }
  res.status(201).json({ media_ids: uploadedIds, message: `${uploadedIds.length} files uploaded` });
});

router.delete('/:projectId/media/:mediaId', requireAdmin, async (req: AuthRequest, res: Response): Promise<void> => {
  const db = await getDb();
  const m = await db.collection('media').findOne({ _id: new ObjectId(req.params.mediaId), project_id: req.params.projectId });
  if (!m) { res.status(404).json({ detail: 'Media not found' }); return; }
  await Promise.all([
    deleteFileFromS3(m.original_key),
    deleteFileFromS3(m.compressed_key),
    deleteFileFromS3(m.thumbnail_key),
    m.watermarked_key ? deleteFileFromS3(m.watermarked_key) : Promise.resolve(),
  ]);
  await db.collection('media').deleteOne({ _id: new ObjectId(req.params.mediaId) });
  res.json({ message: 'Deleted' });
});

router.put('/:projectId/media/reorder', requireAdmin, async (req: AuthRequest, res: Response): Promise<void> => {
  const { media_ids } = req.body;
  if (!Array.isArray(media_ids)) { res.status(400).json({ detail: 'media_ids must be an array' }); return; }
  const db = await getDb();
  await Promise.all(media_ids.map((id: string, idx: number) =>
    db.collection('media').updateOne({ _id: new ObjectId(id), project_id: req.params.projectId }, { $set: { sort_order: idx } })
  ));
  res.json({ message: 'Reordered' });
});

export default router;
